File: CommunityArchiveVectors/create_avatar_mapping.py
"""
Create a simple username -> avatar_url mapping JSON file.
No downloads, just a mapping for the frontend to use.
"""

import json
import os
from pathlib import Path
from supabase import create_client

def main():
    print("Creating username → avatar URL mapping...")

    supabase = create_client(
        os.environ.get("SUPABASE_URL"),
        os.environ.get("SUPABASE_KEY")
    )

    # Get all profiles with avatars and their usernames from tweets
    print("Querying Supabase...")

    # Join all_profile with tweets to get usernames
    query = """
    SELECT DISTINCT ON (p.account_id)
        t.account_display_name as username,
        p.avatar_media_url
    FROM all_profile p
    JOIN tweets t ON p.account_id = t.account_id
    WHERE p.avatar_media_url IS NOT NULL
    """

    response = supabase.rpc('exec_sql', {'query': query}).execute()

    # Alternative approach if RPC doesn't work: fetch separately and join in Python
    print("Fetching profiles...")
    profiles_response = supabase.table("all_profile")\
        .select("account_id, avatar_media_url")\
        .not_("avatar_media_url", "is", None)\
        .execute()

    profiles = {p['account_id']: p['avatar_media_url'] for p in profiles_response.data}
    account_ids = list(profiles.keys())

    print(f"Found {len(profiles)} profiles with avatars")
    print("Fetching usernames...")

    mapping = {}
    batch_size = 1000

    for i in range(0, len(account_ids), batch_size):
        batch = account_ids[i:i + batch_size]

        tweets_response = supabase.table("tweets")\
            .select("account_id, account_display_name")\
            .in_("account_id", batch)\
            .limit(len(batch))\
            .execute()

        for tweet in tweets_response.data:
            account_id = tweet['account_id']
            username = tweet.get('account_display_name')

            if username and account_id in profiles and username not in mapping:
                # Upgrade to higher quality image
                url = profiles[account_id].replace('_normal', '_400x400')
                mapping[username] = url

        print(f"  Processed {min(i + batch_size, len(account_ids))}/{len(account_ids)}...")

    # Save mapping
    output_file = Path("frontend/public/avatar_mapping.json")
    with open(output_file, 'w') as f:
        json.dump(mapping, f, indent=2)

    print(f"\n✓ Created {output_file}")
    print(f"  {len(mapping)} username → avatar URL mappings")
    print(f"\nFrontend usage: const avatar = avatarMap[username];")

if __name__ == "__main__":
    main()

File: CommunityArchiveVectors/create_avatar_mapping.js
/**
 * Create username → avatar URL mapping from Supabase
 * No downloads needed - just a mapping file for the frontend
 */

const { createClient } = require('@supabase/supabase-js');
const fs = require('fs');
const path = require('path');

const supabase = createClient(
  process.env.SUPABASE_URL,
  process.env.SUPABASE_KEY
);

async function main() {
  console.log('Creating username → avatar URL mapping...\n');

  // Get all profiles with avatars
  console.log('Fetching profiles with avatars...');
  const { data: profiles, error: profileError } = await supabase
    .from('all_profile')
    .select('account_id, avatar_media_url')
    .not('avatar_media_url', 'is', null);

  if (profileError) {
    console.error('Error fetching profiles:', profileError);
    return;
  }

  console.log(`Found ${profiles.length} profiles with avatars`);

  // Create account_id → avatar_url map
  const avatarMap = {};
  profiles.forEach(p => {
    // Upgrade to higher quality (_400x400 instead of _normal)
    avatarMap[p.account_id] = p.avatar_media_url.replace('_normal', '_400x400');
  });

  // Get usernames for each account_id in batches
  console.log('Fetching usernames...');
  const accountIds = Object.keys(avatarMap);
  const usernameMapping = {};
  const batchSize = 1000;

  for (let i = 0; i < accountIds.length; i += batchSize) {
    const batch = accountIds.slice(i, i + batchSize);

    const { data: accounts, error: accountsError } = await supabase
      .from('all_account')
      .select('account_id, username')
      .in('account_id', batch);

    if (accountsError) {
      console.error('Error fetching accounts:', accountsError);
      continue;
    }

    accounts.forEach(a => {
      const username = a.username;
      const accountId = a.account_id;
      if (username && accountId in avatarMap) {
        usernameMapping[username] = avatarMap[accountId];
      }
    });

    console.log(`  Processed ${Math.min(i + batchSize, accountIds.length)}/${accountIds.length}...`);
  }

  // Save mapping
  const outputFile = path.join(__dirname, 'frontend/public/avatar_mapping.json');
  fs.writeFileSync(outputFile, JSON.stringify(usernameMapping, null, 2));

  console.log(`\n✓ Created ${outputFile}`);
  console.log(`  ${Object.keys(usernameMapping).length} username → avatar URL mappings`);
  console.log(`\nFrontend usage:`);
  console.log(`  const avatar = avatarMap[username];`);
  console.log(`  <img src={avatar} />`);
}

main().catch(console.error);
